mixnet/client/src/receiver.rs

```diff
@@ -30,26 +42,42 @@ impl Receiver {
         let Ok(socket) = TcpStream::connect(self.node_address).await else {
             return Err(MixnetClientError::MixnetNodeConnectError);
         };
+        let ack_cache = self.ack_cache.clone();
 
         Ok(Self::message_stream(Box::pin(Self::fragment_stream(
-            socket,
+            socket, ack_cache,
         ))))
     }
 
     fn fragment_stream(
         socket: TcpStream,
+        ack_cache: Arc<Mutex<HashMap<SocketAddr, HashSet<PacketId>>>>,
     ) -> impl Stream<Item = Result<Fragment, MixnetClientError>> + Send + 'static {
-        stream::unfold(socket, |mut socket| async move {
-            let Ok(body) = Body::read(&mut socket).await else {
-                // TODO: Maybe this is a hard error and the stream is corrupted? In that case stop the stream
-                return Some((Err(MixnetClientError::MixnetNodeStreamClosed), socket));
-            };
-
-            match body {
-                Body::SphinxPacket(_) => {
-                    Some((Err(MixnetClientError::UnexpectedStreamBody), socket))
+        stream::unfold(socket, move |mut socket| {
+            let cache = ack_cache.clone();
+            async move {
+                let Ok(body) = Body::read(&mut socket).await else {
+                    // TODO: Maybe this is a hard error and the stream is corrupted? In that case stop the stream
+                    return Some((Err(MixnetClientError::MixnetNodeStreamClosed), socket));
+                };
+
+                match body {
+                    Body::SphinxPacket(_) => {
+                        Some((Err(MixnetClientError::UnexpectedStreamBody), socket))
+                    }
+                    Body::FinalPayload(payload) => {
+                        Some((Self::fragment_from_payload(payload), socket))
+                    }
+                    // Client should not receive AckResponse
+                    Body::AckResponse(resp) => {
+                        let mut mu = cache.lock().await;
```

Comment:
I think the sender and the receiver should share the same ack cache, so once the sender sends a message to target A, then we add a message-id to the cache, and when the receiver gets the ack back from the target A, we remove this message-id, so the sender will not retry send anymore. Please correct me if I am wrong.

Reply:
It depends on how you designed the ack protocol, and that's why I would like to see a description of that first. 
Anyway, yes, in principle it's roughly how an ack would work, but it's not clear to me why would a receiver (which is on a different machine from the sender) need to store acks, is it to avoid duplicate acks? If so, why do you think that is important?

Reply:
Yeah, it is my bad, and sorry for missing the description. Let me just explain it here. According to your comment, I think I misunderstood the `SenderReceiver` meaning. I'm not aware that the receiver will be deployed separately, unfortunately. If in this situation, ack cache here is indeed not important. Thanks for pointing it out! It is more clear now, that the sender needs to handle all of the retry logic.

- The flow for client sender:

  - Happy path:

    ```mermaid
    graph TB
        A[Client Sender] -->|packet| B[mixnode]
        B -->|ACK| A
    ```

  - Unhappy path:

    ```mermaid
    graph TB
        A[Client Sender] -->|packet| B[mixnode]
        B -->|no response| A
        A -->|retry packet| B
        B -->|no response| A
        A -->|retry packet| B
        B -->|ACK| A
    ```

- The flow for the mixnode forwarding:

  - Happy path
    ```mermaid
    graph TB
        A[Node A] -->|forward packet| B[Node B]
        B -->|ACK| A
    ```

  - Unhappy path:

    ```mermaid
    graph TB
        A[Node A] -->|forward packet| B[Node B]
        A -->|retry forward packet| B
        B -->|no response| A
        A -->|retry forward packet| B
        B -->|no response| A
        A -->|retry forward packet| B
        B -->|ACK| A
    ```

Reply:
So this solution is for a hop-by-hop individual retry mechanism? We use TCP, that's already done by the underlaying layer

Reply:
Ah, that makes sense, so we only need to refresh the connection if we find a closed connection and send the message through the new connection again.

Reply:
If no errors are reported we can assume everything was delivered and a closed connection is not a problem. We should only retry for transient errors returned by the write call on the tcp stream.

Btw, since we now broadcast everything on libp2p, a cheap end2end verification of successuful delivery would be to look for the message there. Not sure we are interested in this atm

Reply:
> Btw, since we now broadcast everything on libp2p, a cheap end2end verification of successuful delivery would be to look for the message there. Not sure we are interested in this atm

I don't think so. Imo we do not need to know if the message arrived or not. We just need to know message was sent through the tcp stream successfuly.

Reply:
Yes. The same discussions are in https://github.com/logos-co/nomos-node/pull/386#discussion_r1321093763 as well.